src/shell/file_reader.cpp

```diff
@@ -18,69 +18,52 @@
  *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
  */
 
-#include <cassert>
-#include <utility>
-
 #include "file_reader.h"
 
 std::optional<FileReader> FileReader::GetFileReader(const std::string& filename)
 {
+	auto fullname = DOS_Canonicalize(filename.c_str());
+	
 	std::uint16_t handle = {};
-	if (!DOS_OpenFile(filename.c_str(), (DOS_NOT_INHERIT | OPEN_READ), &handle)) {
+	if (!DOS_OpenFile(fullname.c_str(), (DOS_NOT_INHERIT | OPEN_READ), &handle)) {
 		return {};
 	}
-	return FileReader(handle);
+	return FileReader(std::move(fullname));
 }
 
-FileReader::FileReader(const uint16_t file_handle) : handle(file_handle) {}
-
-FileReader::FileReader(FileReader&& other) noexcept
-        : handle(std::exchange(other.handle, {}))
+FileReader::FileReader(std::string filename)
+        : filename(std::move(filename)),
+          cursor(0)
 {}
 
-FileReader& FileReader::operator=(FileReader&& other) noexcept
+std::string FileReader::Read()
```

Comment:
Shouldn't this return an std::optional\<std::string\>? This way, it's easy to distinguish between a failed read (p.e. if the file doesn't open) and an empty line.

And maybe also a way to see what has failed. There are numerous ways I/O can fail, so I think we should be graceful around that subject.

Reply:
Good point @japsmits.